validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/ResponseHandler.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.remote.apiclient;
+
+import static tech.pegasys.teku.infrastructure.http.HttpStatusCodes.SC_ACCEPTED;
+import static tech.pegasys.teku.infrastructure.http.HttpStatusCodes.SC_BAD_REQUEST;
+import static tech.pegasys.teku.infrastructure.http.HttpStatusCodes.SC_NO_CONTENT;
+import static tech.pegasys.teku.infrastructure.http.HttpStatusCodes.SC_OK;
+import static tech.pegasys.teku.infrastructure.http.HttpStatusCodes.SC_SERVICE_UNAVAILABLE;
+import static tech.pegasys.teku.infrastructure.http.HttpStatusCodes.SC_TOO_MANY_REQUESTS;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import okhttp3.Request;
+import okhttp3.Response;
+import tech.pegasys.teku.provider.JsonProvider;
+
+public class ResponseHandler<T> {
+  private final Map<Integer, Handler<T>> handlers = new HashMap<>();
+  private final JsonProvider jsonProvider;
+  private final Class<T> responseClass;
+
+  public ResponseHandler(final JsonProvider jsonProvider, final Class<T> responseClass) {
+    this.jsonProvider = jsonProvider;
+    this.responseClass = responseClass;
+    withHandler(SC_OK, this::defaultOkHandler);
+    withHandler(SC_ACCEPTED, this::noValueHandler);
+    withHandler(SC_NO_CONTENT, this::noValueHandler);
+    withHandler(SC_SERVICE_UNAVAILABLE, this::noValueHandler);
+    withHandler(SC_BAD_REQUEST, this::defaultBadRequestHandler);
+    withHandler(SC_TOO_MANY_REQUESTS, this::defaultTooManyRequestsHandler);
```

Comment:
i would have thought putting in SC_NOT_FOUND here as noValueHandler would have been adequate? it'll be a common use case potentially...

Reply:
No that's what we moved away from because it silently ignored the fact that Infura hadn't enabled any of the new APIs.  Most APIs shouldn't return a 404 or should be considered an error worth notifying the user about if they do.

Reply:
ok, every api that returns a single element will be in that list, they all return 404, which is 'correct' behaviour, so the consumers will need to individually decide how to handle whatever they use. that's fine.